### PR TITLE
Adding TabletServerLocks utility to accumulo admin command

### PR DESCRIPTION
**Pull request by cshannon**
Issue #2807

### CODE REVIEW

**Comment by milleruntime** [comment]
I think this one should go under Admin command as well.

**Comment by cshannon** [comment]
@milleruntime - This has been updated to move to the Admin command

File: server/base/src/main/java/org/apache/accumulo/server/util/Admin.java [hunk 1]
```diff
@@ -217,6 +226,9 @@ public void execute(final String[] args) {
     RandomizeVolumesCommand randomizeVolumesOpts = new RandomizeVolumesCommand();
     cl.addCommand("randomizeVolumes", randomizeVolumesOpts);
 
+    TabletServerLocksCommand tServerLocksOpts = new TabletServerLocksCommand();
+    cl.addCommand("tServerLocks", tServerLocksOpts);
```
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/util/Admin.java:230]: Could call the command `locks`. That is a lot shorter and easier to type.

File: server/base/src/main/java/org/apache/accumulo/server/util/Admin.java [hunk 2]
```diff
@@ -162,6 +162,15 @@ static class RandomizeVolumesCommand {
     String tableName = null;
   }
 
+  @Parameters(commandDescription = "List or delete Tablet Server locks")
+  static class TabletServerLocksCommand {
+    @Parameter(names = "-list", description = "print the tablet server locks")
+    boolean list = false;
+
+    @Parameter(names = "-delete", description = "specific a tablet server lock to delete")
```
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/util/Admin.java:170]:
```suggestion
    @Parameter(names = "-delete", description = "specify a tablet server lock to delete")
```

**Review by milleruntime — CHANGES_REQUESTED** [review]
Looks good so far but I think the command could be improved a little bit. The default behavior (if the user doesn't provide a parameter) should just be to list the locks. And the command `tServerLocks` is a bit awkward to type. It could be shortened to something like `locks`.

File: server/base/src/main/java/org/apache/accumulo/server/util/TabletServerLocks.java [hunk 3]
```diff
@@ -41,38 +41,50 @@ static class Opts extends Help {
...
+      }
+      ServiceLock.deleteLock(zoo, ServiceLock.path(tserverPath + "/" + lock));
+    } else {
+      printUsage(commandName);
+    }
+  }
+
+  public static void main(String[] args) throws Exception {
 
+    try (var context = new ServerContext(SiteConfiguration.auto())) {
```
Comment (Manno15) [inline server/base/src/main/java/org/apache/accumulo/server/util/TabletServerLocks.java:77]: Should this command also aim to refactor the use of SiteConfiguration like in issue #2155?

**Comment by cshannon** [comment]
> Looks good so far but I think the command could be improved a little bit. The default behavior (if the user doesn't provide a parameter) should just be to list the locks. And the command `tServerLocks` is a bit awkward to type. It could be shortened to something like `locks`.

Sounds good, I'll make these changes

File: server/base/src/main/java/org/apache/accumulo/server/util/TabletServerLocks.java (hunk 3 above)
Comment (milleruntime) [reply server/base/src/main/java/org/apache/accumulo/server/util/TabletServerLocks.java:73]: Yeah that is a good point. This one may need more refactoring to allow for that though. I think the main method should probably be dropped too if its going to be run through the Admin command. Then there should be a way to get the `ServerContext` from Admin.

**Comment by cshannon** [comment]
@milleruntime - Can this issue and also #2814 be merged or do we want to try and get rid of SiteConfiguration now? I think it makes more sense to merge as is and get rid of SiteConfiguration as a follow on task since more refactoring and work would probably be needed.

**Comment by Manno15** [comment]
> Can this issue and also https://github.com/apache/accumulo/pull/2814 be merged or do we want to try and get rid of SiteConfiguration now? I think it makes more sense to merge as is and get rid of SiteConfiguration as a follow on task since more refactoring and work would probably be needed.

@milleruntime I am personally in favor of merging this as is and discussing whether or not we want to go through and remove siteConfiguration from each of these commands that use them.

**Comment by ctubbsii** [comment]
Without taking a deeper look, I'm not sure what this utility is specifically doing, so I can't speak to the benefits of where its entry point should be. However, one comment I'll make about SiteConfiguration is: the point of removing SiteConfiguration was to ensure that we don't have a dependency, or in general, make any assumptions about where the shell was being run, relative to where server processes are deployed. For example, there's no reason to assume that the server configuration file (the site configuration), with the `instance.secret` for writing to protected nodes in ZooKeeper, is available on the class path for the shell. The shell is a client application, and should not be assumed to have access to the server's configuration file at all.

The same is not true for utilities specifically written to assist system administrators perform maintenance on Accumulo itself on the server side. Many such utilities are required to be run on the server itself, so it is reasonable for them to assume they have access to the server configuration on their class path.

Let's be careful not to lose site of the point of getting SiteConfiguration out of the shell, and forcing all these utilities to run client side, and then needing to bake in convoluted hacks to get them to run off the server. It is perfectly acceptable for server-side utilities to just assume access to server-side configuration and class path.

**Comment by milleruntime** [comment]
Thanks @ctubbsii for keeping things in perspective. I think this server utility fits well under the `Admin` command but I think the code could be cleaned up. I am going to take a look at it.

**Review by milleruntime — CHANGES_REQUESTED** [review]
I made a PR to your branch with some ideas. I still couldn't get the usage to print though so we might just want to drop that method as well.

**Comment by cshannon** [comment]
@milleruntime - In your update you removed the main method, do you still want to do that now or do it later like the others?

**Comment by milleruntime** [comment]
> @milleruntime - In your update you removed the main method, do you still want to do that now or do it later like the others?

You can just ignore that update to your PR for now.

File: server/base/src/main/java/org/apache/accumulo/server/util/TabletServerLocks.java (hunk 3 above)
Comment (milleruntime) [reply server/base/src/main/java/org/apache/accumulo/server/util/TabletServerLocks.java:73]: See Christopher's [comment](https://github.com/apache/accumulo/pull/2817#issuecomment-1208327394) about the difference between running in the Shell vs Admin.

**Review by milleruntime — APPROVED** [review]

**Comment by milleruntime** [comment]
OK @cshannon I think I got all your PRs that had additions to the Admin command. Please just have 1 follow up PR with all the improvements.

**Pull request merged**